### ISSUE TITLE
pyilper 1.8.8 crash on start

### ISSUE THREAD

**Issue by pcscote** [bug400/pyilper#6]
pyilper 1.8.8 systematically crash on start

uname -a
```
Darwin xyz.local 23.3.0 Darwin Kernel Version 23.3.0: 
Wed Dec 20 21:28:58 PST 2023; root:xnu-10002.81.5~7/RELEASE_X86_64 x86_64
```

sw_vers
```
ProductName:		macOS
ProductVersion:		14.3.1
BuildVersion:		23D60
```

system_profiler SPSoftwareDataType
```
    System Software Overview:

      System Version: macOS 14.3.1 (23D60)
      Kernel Version: Darwin 23.3.0
      Boot Volume: Macintosh
      Boot Mode: Normal
      Computer Name: xyz
      User Name: me
      Secure Virtual Memory: Enabled
      System Integrity Protection: Enabled
      Time since boot: 9 days, 11 hours, 48 minutes

```

which python3
`/Users/me/py311/bin/python3`

python3 --version
`Python 3.11.7`

python3 -m pip list
```
Package            Version
------------------ -------
pip                24.0
pyilper            1.8.8
pyserial           3.5
PySide6            6.6.2
PySide6_Addons     6.6.2
PySide6_Essentials 6.6.2
setuptools         69.1.1
shiboken6          6.6.2
```

python3 -m pip cache purge 
```
WARNING: No matching packages
Files removed: 0
```

which pyilper
`/Users/me/py311/bin/pyilper`

pyilper                   
```
Traceback (most recent call last):
  File "signature_bootstrap.py", line 77, in bootstrap
  File "signature_bootstrap.py", line 93, in find_incarnated_files
  File "/usr/local/Cellar/python@3.11/3.11.7_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/pathlib.py", line 871, in __new__
    self = cls._from_parts(args)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/python@3.11/3.11.7_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/pathlib.py", line 509, in _from_parts
    drv, root, parts = self._parse_args(args)
                       ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/python@3.11/3.11.7_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/pathlib.py", line 493, in _parse_args
    a = os.fspath(a)
        ^^^^^^^^^^^^
TypeError: expected str, bytes or os.PathLike object, not NoneType
Fatal Python error: could not initialize part 2
Python runtime state: initialized

Current thread 0x00007ff851d207c0 (most recent call first):
  File "<frozen importlib._bootstrap>", line 241 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 1233 in create_module
  File "<frozen importlib._bootstrap>", line 573 in module_from_spec
  File "<frozen importlib._bootstrap>", line 676 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1147 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1176 in _find_and_load
  File "<frozen importlib._bootstrap>", line 241 in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1232 in _handle_fromlist
  File "/Users/me/py311/lib/python3.11/site-packages/PySide6/__init__.py", line 64 in _setupQtDirectories
  File "/Users/me/py311/lib/python3.11/site-packages/PySide6/__init__.py", line 124 in <module>
  File "<frozen importlib._bootstrap>", line 241 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 940 in exec_module
  File "<frozen importlib._bootstrap>", line 690 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1147 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1176 in _find_and_load
  File "<frozen importlib._bootstrap>", line 241 in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1126 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1176 in _find_and_load
  File "/Users/me/py311/lib/python3.11/site-packages/pyilper/pilcore.py", line 273 in <module>
  File "<frozen importlib._bootstrap>", line 241 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 940 in exec_module
  File "<frozen importlib._bootstrap>", line 690 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1147 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1176 in _find_and_load
  File "/Users/me/py311/lib/python3.11/site-packages/pyilper/pyilpermain.py", line 199 in <module>
  File "<frozen importlib._bootstrap>", line 241 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 940 in exec_module
  File "<frozen importlib._bootstrap>", line 690 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1147 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1176 in _find_and_load
  File "/Users/me/py311/lib/python3.11/site-packages/pyilper/__init__.py", line 1 in <module>
  File "<frozen importlib._bootstrap>", line 241 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 940 in exec_module
  File "<frozen importlib._bootstrap>", line 690 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1147 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1176 in _find_and_load
  File "/Users/me/py311/bin/pyilper", line 5 in <module>

Extension modules: xxsubtype (total: 1)
zsh: abort      pyilper
```

**Comment by pcscote** [comment]
I have created another virtual environment and it is now working but it is very flimsy.
As an example, if I start pyilper without a PILBox connected, the dropdown menu does not works.

**Comment by bug400** [comment]
I could reproduce the crash of your first post by updating Pyside6 from 6.6.1 to 6.6.2 in my current test environment. There are issues with this PySide6 version (see https://bugreports.qt.io/browse/PYSIDE-2616).

After installing a new clean environment with PySide 6.6.2 pyILPER ran fine.

Perhaps it helps to set up a clean new environment with PySide 6.6.1:

     python3 -m venv test661

     source test661/bin/activate

     python -m pip cache purge

     python -m pip install PySide6==6.6.1

     python -m pip install pyilper

     pyilper

**Comment by pcscote** [comment]
Hello Joachim,

I have done these steps:


cd ~
rm -Rf ~/Library/Application\ Support/pyilper
python3 -m venv ppenv
source ppenv/bin/activate
python3 -m pip cache purge
python3 -m pip install PySide6==6.6.1
python3 -m pip install pyilper
pyilper

in pyilper app
a) close first time dialog
b) close "Serial device not configured : Run pyILPER configuration" dialog
c) menu: File - pyILPER configuration

in pyILPER configuration dialog
a) Communication configuration
b) PIL-Box selected
c) Serial Device - "change"button pressed
d) crash!

Qt has caught an exception thrown from an event handler. Throwing
exceptions from an event handler is not supported in Qt.
You must not let any exception whatsoever propagate through Qt code.
Fatal Python error: PyGILState_Release: thread state 0x10ebe5d58 must be current when releasing
Python runtime state: initialized

Current thread 0x00007ff8569387c0 (most recent call first):
  <no Python frame>

Extension modules: xxsubtype, shiboken6.Shiboken, PySide6.QtCore, PySide6.QtGui, PySide6.QtWidgets, PySide6.QtNetwork, PySide6.QtPrintSupport, PySide6.QtWebChannel, PySide6.QtWebEngineCore, PySide6.QtWebEngineWidgets (total: 10)

zsh: abort      pyilper

Best regards,

Sylvain

**Comment by bug400** [comment]
Hello Sylvain,
the error message above is so strange that I assume that there are incompatibilities in the runtime environment you are using. I strongly recommend the Python 3.11 or 3.12 version provided by the python.org website. According to the Python path shown in the error trace of comment 1 I assume that you are using a Homebrew Python version. I admit that the python.org version is bulky and difficult to uninstall (it puts symlinks into /usr/local/) but according to my experience it is compatible with the PySide6 version from PyPi.

Regards
Joachim

**Comment by pcscote** [comment]
Hello Joachim,
Finally some free time to solve this issue.
I have uninstalled brew, python3, some other files and cleanup my zsh config files.
Then I have reinstalled brew, python3 (3.12), some other packages and updated my zsh config files.
Finally, I have recreated my virtual environment (ppenv) and followed your installation instruction.
`Successfully installed PySide6-Addons-6.6.2 PySide6-Essentials-6.6.2 pyilper-1.8.8 pyserial-3.5 pyside6-6.6.2 shiboken6-6.6.2
`
Everything is working as expected now, thanks!
Regards,
Sylvain

**Comment by bug400** [comment]
Hello Sylvain,
Glad to hear this. Do you run pyILPER now with a cleanly installed "brew Python" or a "python.org Python"?
Regards
Joachim

**Comment by pcscote** [comment]
Hello Joachim,

brew Python (Python 3.12.2)

Installation steps that I did: (minus the unrelated packages)
```
cd
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
brew update
brew upgrade
brew install python3
python3 -m venv ppenv
source ppenv/bin/activate
python -m pip install pyilper
```
Then, for each pil-box connected and/or specific project, I start one or multiple pyilper instances
```
pyilper --instance=pb1 &
pyilper --instance=pb2 &
```

For the more common configurations I have created macOS automator scripts.
Like: controller-to-device (c2d),  device-to-controller (d2c), etc.

ex.: "pyilper-d2c" automator application with a "run shell script" action with these lines in it
```
cd
source ppenv/bin/activate
mkdir -p pyilper/d2c
cd pyilper/d2c
pyilper --instance=d2c
```

Best regards,

Sylvain

PS: Thank you very much for having created and for maintaining this very useful piece of software, I use it daily.

PPS: If you have nothing to do and searching for something to fill up your evenings ...  ;-)
I am frequently using more than one PIL-Box in a loop and in some cases, I need to bridge the PIL-Box (IL/USB) to the TCP/IP interface (IL over TCP). I currently use CG ILPilBox bridge software for these ones, but it would be nice if pyILPer had that capability built-in. You are already supporting both communication interfaces, what is missing is the capability to forward IL messages across them.
